experiments/sat/sat_tensornetwork.py

```diff
@@ -0,0 +1,166 @@
+          ) -> Tuple[tensornetwork.TensorNetwork, List[tensornetwork.Edge]]:
+  """Create a 3SAT TensorNetwork of the given 3SAT clauses.
+
+    After full contraction, this network will be a tensor of size (2, 2, ..., 2)
+    with the rank being the same as the number of variables. Each element of the
+    final tensor represents whether the given assignment satisfies all of the
+    clauses. For example, if final_node.get_tensor()[0][1][1] == 1, then the
+    assiment (False, True, True) satisfies all clauses.
+
+
+  Args:
+    clauses: A list of 3 int tuples. Each element in the tuple corresponds to a
+      variable in the clause. If that int is negative, that variable is negated
+      in the clause.
+
+  Returns:
+    net: The 3SAT TensorNetwork.
+    var_edges: The edges for the given variables.
+
+  Raises:
+    ValueError: If any of the clauses have a 0 in them.
+  """
+  for clause in clauses:
+    if 0 in clause:
+      raise ValueError("0's are not allowed in the clauses.")
+  net = tensornetwork.TensorNetwork()
+  var_set = set()
+  for clause in clauses:
+    var_set |= {abs(x) for x in clause}
+  num_vars = max(var_set)
+  var_nodes = []
+  var_edges = []
+
+  # Prepare the variable nodes.
+  for _ in range(num_vars):
+    new_node = net.add_node(np.ones(2, dtype=np.int32))
+    var_nodes.append(new_node)
+    var_edges.append(new_node[0])
+
+  # Create the nodes for each clause
+  for clause in clauses:
+    a, b, c, = clause
+    clause_tensor = np.ones((2, 2, 2), dtype=np.int32)
+    clause_tensor[(-np.sign(a) + 1) // 2, (-np.sign(b) + 1) // 2,
+                  (-np.sign(c) + 1) // 2] = 0
+    clause_node = net.add_node(clause_tensor)
+
+    # Connect the variable to the clause through a copy tensor.
+    for i, var in enumerate(clause):
+      copy_tensor_node = net.add_node(_build_copy_tensor())
+      net.connect(clause_node[i], copy_tensor_node[0])
+      net.connect(var_edges[abs(var) - 1], copy_tensor_node[1])
+      var_edges[abs(var) - 1] = copy_tensor_node[2]
+
+  return net, var_edges
+
+
+def sat_count_tn(clauses: List[Tuple[int, int, int]]
+                ) -> tensornetwork.TensorNetwork:
+  """Create a 3SAT Count TensorNetwork.
+
+  After full contraction, the final node will be the count of all possible
+  solutions to the given 3SAT problem.
+
+  Args:
+    clauses: A list of 3 int tuples. Each element in the tuple corresponds to a
+      variable in the clause. If that int is negative, that variable is negated
+      in the clause.
+
+  Returns:
+    net: The TensorNetwork corresponding to the norm of the 3SAT TensorNetwork
+      generated by the given clauses.
+  """
+  net1, var_edges1 = sat_tn(clauses)
+  net2, var_edges2 = sat_tn(clauses)
+  sat_net = tensornetwork.TensorNetwork.merge_networks([net1, net2])
+  for edge1, edge2 in zip(var_edges1, var_edges2):
+    sat_net.connect(edge1, edge2)
+  return sat_net
+
+
+def contract_badly(net: tensornetwork.TensorNetwork,
```

Comment:
nit: I'd change this name to better reflect what problem the user should expect. Current name suggests that there may be an issue with correctness, which is not the case. The problem relates to performance. Perhaps something like `contract_slow()`?

Reply:
How about I delete it entirely and replace it with naive?

Reply:
Sure, that works, too.